packages/tools/src/utilities/math/polyline/combinePolyline.ts

```diff
@@ -0,0 +1,327 @@
+import getNormal2 from './getNormal2';
+import { glMatrix, vec3 } from 'gl-matrix';
+import getLineSegmentsIntersection from './getLineSegmentsIntersection';
+
+enum PolylinePointType {
+  Vertex,
+  Intersection,
+}
+
+// Position of the point related to the intersection region
+enum PolylinePointPosition {
+  Outside = -1,
+  Edge = 0,
+  Inside = 1,
+}
+
+// Direction from last point to the intersection point to know if it is entering
+// or exiting the intersection region
+enum PolylinePointDirection {
+  Exiting = -1,
+  Unknown = 0,
+  Entering = 1,
+}
+
+type PolylinePoint = {
+  type: PolylinePointType;
+  coordinates: Types.Point2;
+  position?: PolylinePointPosition;
+  visited: boolean;
+  next: PolylinePoint;
+};
+
+type PolylineIntersectionPoint = PolylinePoint & {
+  direction: PolylinePointDirection;
+  cloned?: boolean;
+};
+
+function ensuresNextPointers(polylinePoints: PolylinePoint[]) {
+  // Make sure all nodes point to a valid node
+  for (let i = 0, len = polylinePoints.length; i < len; i++) {
+    const currentPoint = polylinePoints[i];
+
+    if (!currentPoint.next) {
+      currentPoint.next = polylinePoints[i === len - 1 ? 0 : i + 1];
+    }
+  }
+}
+
+function getSourceAndTargetPointsList(
+  targetPolyline: Types.Point2[],
+  sourcePolyline: Types.Point2[]
+) {
+  const targetPolylinePoints: PolylinePoint[] = [];
+  const sourcePolylinePoints: PolylinePoint[] = [];
+  const sourceIntersectionsCache = new Map<
+    number,
+    PolylineIntersectionPoint[]
+  >();
+
+  let intersectionPointDirection = PolylinePointDirection.Entering;
+
+  // Store all vertices and intersection for target contour
+  for (let i = 0, len = targetPolyline.length; i < len; i++) {
+    const p1 = targetPolyline[i];
+    const pointInside = containsPoint(sourcePolyline, p1);
+    const vertexPoint: PolylinePoint = {
+      type: PolylinePointType.Vertex,
+      coordinates: p1,
+      position: pointInside
+        ? PolylinePointPosition.Inside
+        : PolylinePointPosition.Outside,
+      visited: false,
+      next: null,
+    };
+
+    targetPolylinePoints.push(vertexPoint);
+
+    const q1 = targetPolyline[i === len - 1 ? 0 : i + 1];
+    const intersectionsInfo = getLineSegmentIntersectionsIndexes(
+      sourcePolyline,
+      p1,
+      q1
+    ).map((intersectedLineSegment) => {
+      const sourceLineSegmentId: number = intersectedLineSegment[0];
+      const p2 = sourcePolyline[intersectedLineSegment[0]];
+      const q2 = sourcePolyline[intersectedLineSegment[1]];
+
+      // lineSegment.intersectLine returns `undefined` when the intersection
+      // is at one of the line vertices.
+      // Examples:
+      //   - [(0, 0), (1, 1)] x [(1, 1), (1, 2)]
+      //   - [(0, 1), (2, 1)] x [(1, 1), (1, 2)]
+      const intersectionCoordinate = getLineSegmentsIntersection(
+        p1,
+        q1,
+        p2,
+        q2
+      ) as Types.Point2;
+
+      const targetStartPointDistSquared = math.point.distanceToPointSquared(
```

Comment:
can't we get away with distanceToPoint (not squared) to avoid squaring operation, since you need it for sort only the non-squared version is as good as the squared version

Reply:
I didn't get. Are you saying to call `distanceToPoint` instead of `distanceToPointSquared`? But `distanceToPoint` calls `distanceToPointSquared` internally and it also takes the square root that is one of the most costly operator.

Time to process 10M numbers:
 - (n * n) = 161ms
 - (n ** 2) = 199ms
 - Math.pow(n, 2) = 29529ms
 - Math.sqrt(n) = 31946ms

Reply:
What I mean is you only need this distance for sorting right? Then you don't need squared or squared root, you can just use the products and sum

Reply:
as we discussed there are no changes to be done.